{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# **What causes antibiotic resistance?** \n",
    "Project 1\\\n",
    "*Lab Journal by Ustin Zolotikov*\n",
    "\n",
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 1\n",
    "__*22/10/2023*__\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir project_1\n",
    "!cd project_1\n",
    "!mamba create -n project_1 -c bioconda trimmomatic fastqc bwa samtools igv varscan snpeff\n",
    "!mamba activate project_1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made project dir and env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-22 21:08:16--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 165.112.9.228, 165.112.9.230, 2607:f220:41e:250::12, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|165.112.9.228|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1379902 (1,3M) [application/x-gzip]\n",
      "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’\n",
      "\n",
      "GCF_000005845.2_ASM 100%[===================>]   1,32M   908KB/s    in 1,5s    \n",
      "\n",
      "2023-10-22 21:08:19 (908 KB/s) - ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’ saved [1379902/1379902]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downloaded a reference sequence of the parental (unevolved, not resistant to antibiotics) E. coli strain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-22 21:08:57--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 165.112.9.228, 165.112.9.230, 2607:f220:41e:250::12, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|165.112.9.228|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 433547 (423K) [application/x-gzip]\n",
      "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.gff.gz’\n",
      "\n",
      "GCF_000005845.2_ASM 100%[===================>] 423,39K   446KB/s    in 0,9s    \n",
      "\n",
      "2023-10-22 21:08:59 (446 KB/s) - ‘GCF_000005845.2_ASM584v2_genomic.gff.gz’ saved [433547/433547]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dowloaded an annotation of  the reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-22 21:12:11--  http://wget/\n",
      "Resolving wget (wget)... failed: Name or service not known.\n",
      "wget: unable to resolve host address ‘wget’\n",
      "--2023-10-22 21:12:11--  https://figshare.com/ndownloader/files/23769689\n",
      "Resolving figshare.com (figshare.com)... 108.128.65.255, 18.203.60.14, 2a05:d018:1f4:d003:b4f6:8fdd:a5f7:7369, ...\n",
      "Connecting to figshare.com (figshare.com)|108.128.65.255|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20231022/eu-west-1/s3/aws4_request&X-Amz-Date=20231022T180743Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=984252d9649042af4636fa07de686ca3be04f5ff5bba99ed21033e8964a2fb11 [following]\n",
      "--2023-10-22 21:12:12--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20231022/eu-west-1/s3/aws4_request&X-Amz-Date=20231022T180743Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=984252d9649042af4636fa07de686ca3be04f5ff5bba99ed21033e8964a2fb11\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.112.139, 52.92.0.64, 52.218.108.51, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.112.139|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 44390940 (42M) [application/gzip]\n",
      "Saving to: ‘23769689’\n",
      "\n",
      "23769689            100%[===================>]  42,33M  99,7KB/s    in 10m 12s \n",
      "\n",
      "2023-10-22 21:22:26 (70,8 KB/s) - ‘23769689’ saved [44390940/44390940]\n",
      "\n",
      "FINISHED --2023-10-22 21:22:26--\n",
      "Total wall clock time: 10m 15s\n",
      "Downloaded: 1 files, 42M in 10m 12s (70,8 KB/s)\n"
     ]
    }
   ],
   "source": [
    "!wget wget https://figshare.com/ndownloader/files/23769689"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-22 21:22:29--  https://figshare.com/ndownloader/files/23769692\n",
      "Resolving figshare.com (figshare.com)... 18.203.60.14, 108.128.65.255, 2a05:d018:1f4:d003:b4f6:8fdd:a5f7:7369, ...\n",
      "Connecting to figshare.com (figshare.com)|18.203.60.14|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769692/amp_res_2.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20231022/eu-west-1/s3/aws4_request&X-Amz-Date=20231022T181801Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=911b116cf8220820a4a88224285a3dc269bf0594cf6ed2d6bd2072b433b3d426 [following]\n",
      "--2023-10-22 21:22:30--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769692/amp_res_2.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20231022/eu-west-1/s3/aws4_request&X-Amz-Date=20231022T181801Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=911b116cf8220820a4a88224285a3dc269bf0594cf6ed2d6bd2072b433b3d426\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.92.36.0, 52.218.36.234, 52.218.108.227, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.92.36.0|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 44375379 (42M) [application/gzip]\n",
      "Saving to: ‘23769692’\n",
      "\n",
      "23769692            100%[===================>]  42,32M   202KB/s    in 4m 59s  \n",
      "\n",
      "2023-10-22 21:27:30 (145 KB/s) - ‘23769692’ saved [44375379/44375379]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://figshare.com/ndownloader/files/23769692"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download of forward and reverse reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip 237*\n",
    "!ln -s 23769689 R1.fastq\n",
    "!ln -s 23769692 R2.fastq"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Making symbolic links for read 1 and read 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!wc -l R1.fastq # 1823504/4 = 455876\n",
    "!wc -l R2.fastq # 1823504/4 = 455876"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Counted number of reads in each file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir fastqc\n",
    "fastqc R*.fastq -o fastqc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fastqc analysis for [read 1](./fastqc/R1_fastqc.html) and [read 2]((./fastqc/R2_fastqc.html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir trimmed\n",
    "trimmomatic PE -threads 4 -phred33 R1.fastq R2.fastq \\\n",
    "    trimmed/R1.trimmed.fastq trimmed/R1_un.trimmed.fastq \\\n",
    "    trimmed/R2.trimmed.fastq trimmed/R2_un.trimmed.fastq \\\n",
    "    LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trimmed the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1785036 trimmed/R1.trimmed.fastq\n",
      "1785036 trimmed/R2.trimmed.fastq\n"
     ]
    }
   ],
   "source": [
    "!wc -l trimmed/R1.trimmed.fastq # 1785036/4 = 446259\n",
    "!wc -l trimmed/R2.trimmed.fastq # 1785036/4 = 446259"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Counted reads in trimmed fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir fastqc_trimmed\n",
    "fastqc trimmed/* -o fastqc_trimmed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made fastqc analysis on trimmed reads"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 2\n",
    "*__23/10/2023__*\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘trimmed_30’: File exists\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 4 -phred33 R1.fastq R2.fastq trimmed_30/R1.trimmed.fastq trimmed_30/R1_un.trimmed.fastq trimmed_30/R2.trimmed.fastq trimmed_30/R2_un.trimmed.fastq LEADING:30 TRAILING:30 SLIDINGWINDOW:10:30 MINLEN:20\n",
      "Input Read Pairs: 455876 Both Surviving: 376340 (82.55%) Forward Only Surviving: 33836 (7.42%) Reverse Only Surviving: 25307 (5.55%) Dropped: 20393 (4.47%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir trimmed_30\n",
    "trimmomatic PE -threads 4 -phred33 R1.fastq R2.fastq \\\n",
    "    trimmed_30/R1.trimmed.fastq trimmed_30/R1_un.trimmed.fastq \\\n",
    "    trimmed_30/R2.trimmed.fastq trimmed_30/R2_un.trimmed.fastq \\\n",
    "    LEADING:30 TRAILING:30 SLIDINGWINDOW:10:30 MINLEN:20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trimmed with quality 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "text/plain\n",
      "text/plain\n",
      "text/plain\n",
      "text/plain\n",
      "Started analysis of R1.trimmed.fastq\n",
      "Approx 5% complete for R1.trimmed.fastq\n",
      "Approx 10% complete for R1.trimmed.fastq\n",
      "Approx 15% complete for R1.trimmed.fastq\n",
      "Approx 20% complete for R1.trimmed.fastq\n",
      "Approx 25% complete for R1.trimmed.fastq\n",
      "Approx 30% complete for R1.trimmed.fastq\n",
      "Approx 35% complete for R1.trimmed.fastq\n",
      "Approx 40% complete for R1.trimmed.fastq\n",
      "Approx 45% complete for R1.trimmed.fastq\n",
      "Approx 50% complete for R1.trimmed.fastq\n",
      "Approx 55% complete for R1.trimmed.fastq\n",
      "Approx 60% complete for R1.trimmed.fastq\n",
      "Approx 65% complete for R1.trimmed.fastq\n",
      "Approx 70% complete for R1.trimmed.fastq\n",
      "Approx 75% complete for R1.trimmed.fastq\n",
      "Approx 80% complete for R1.trimmed.fastq\n",
      "Approx 85% complete for R1.trimmed.fastq\n",
      "Approx 90% complete for R1.trimmed.fastq\n",
      "Approx 95% complete for R1.trimmed.fastq\n",
      "Analysis complete for R1.trimmed.fastq\n",
      "Started analysis of R1_un.trimmed.fastq\n",
      "Approx 5% complete for R1_un.trimmed.fastq\n",
      "Approx 10% complete for R1_un.trimmed.fastq\n",
      "Approx 15% complete for R1_un.trimmed.fastq\n",
      "Approx 20% complete for R1_un.trimmed.fastq\n",
      "Approx 25% complete for R1_un.trimmed.fastq\n",
      "Approx 30% complete for R1_un.trimmed.fastq\n",
      "Approx 35% complete for R1_un.trimmed.fastq\n",
      "Approx 40% complete for R1_un.trimmed.fastq\n",
      "Approx 45% complete for R1_un.trimmed.fastq\n",
      "Approx 50% complete for R1_un.trimmed.fastq\n",
      "Approx 55% complete for R1_un.trimmed.fastq\n",
      "Approx 60% complete for R1_un.trimmed.fastq\n",
      "Approx 65% complete for R1_un.trimmed.fastq\n",
      "Approx 70% complete for R1_un.trimmed.fastq\n",
      "Approx 75% complete for R1_un.trimmed.fastq\n",
      "Approx 80% complete for R1_un.trimmed.fastq\n",
      "Approx 85% complete for R1_un.trimmed.fastq\n",
      "Approx 90% complete for R1_un.trimmed.fastq\n",
      "Approx 95% complete for R1_un.trimmed.fastq\n",
      "Analysis complete for R1_un.trimmed.fastq\n",
      "Started analysis of R2.trimmed.fastq\n",
      "Approx 5% complete for R2.trimmed.fastq\n",
      "Approx 10% complete for R2.trimmed.fastq\n",
      "Approx 15% complete for R2.trimmed.fastq\n",
      "Approx 20% complete for R2.trimmed.fastq\n",
      "Approx 25% complete for R2.trimmed.fastq\n",
      "Approx 30% complete for R2.trimmed.fastq\n",
      "Approx 35% complete for R2.trimmed.fastq\n",
      "Approx 40% complete for R2.trimmed.fastq\n",
      "Approx 45% complete for R2.trimmed.fastq\n",
      "Approx 50% complete for R2.trimmed.fastq\n",
      "Approx 55% complete for R2.trimmed.fastq\n",
      "Approx 60% complete for R2.trimmed.fastq\n",
      "Approx 65% complete for R2.trimmed.fastq\n",
      "Approx 70% complete for R2.trimmed.fastq\n",
      "Approx 75% complete for R2.trimmed.fastq\n",
      "Approx 80% complete for R2.trimmed.fastq\n",
      "Approx 85% complete for R2.trimmed.fastq\n",
      "Approx 90% complete for R2.trimmed.fastq\n",
      "Approx 95% complete for R2.trimmed.fastq\n",
      "Analysis complete for R2.trimmed.fastq\n",
      "Started analysis of R2_un.trimmed.fastq\n",
      "Approx 5% complete for R2_un.trimmed.fastq\n",
      "Approx 10% complete for R2_un.trimmed.fastq\n",
      "Approx 15% complete for R2_un.trimmed.fastq\n",
      "Approx 20% complete for R2_un.trimmed.fastq\n",
      "Approx 25% complete for R2_un.trimmed.fastq\n",
      "Approx 30% complete for R2_un.trimmed.fastq\n",
      "Approx 35% complete for R2_un.trimmed.fastq\n",
      "Approx 40% complete for R2_un.trimmed.fastq\n",
      "Approx 45% complete for R2_un.trimmed.fastq\n",
      "Approx 50% complete for R2_un.trimmed.fastq\n",
      "Approx 55% complete for R2_un.trimmed.fastq\n",
      "Approx 60% complete for R2_un.trimmed.fastq\n",
      "Approx 65% complete for R2_un.trimmed.fastq\n",
      "Approx 70% complete for R2_un.trimmed.fastq\n",
      "Approx 75% complete for R2_un.trimmed.fastq\n",
      "Approx 80% complete for R2_un.trimmed.fastq\n",
      "Approx 85% complete for R2_un.trimmed.fastq\n",
      "Approx 90% complete for R2_un.trimmed.fastq\n",
      "Approx 95% complete for R2_un.trimmed.fastq\n",
      "Analysis complete for R2_un.trimmed.fastq\n"
     ]
    }
   ],
   "source": [
    "!mkdir fastqc_trimmed_30\n",
    "!fastqc trimmed_30/* -o fastqc_trimmed_30"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made fastqc analysis for new trimmed reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.08 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 3.65 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.04 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.06 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 1.22 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index ref/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
      "[main] Real time: 10.370 sec; CPU: 5.308 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa index ref/GCF*fna.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made index for the reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 106336 sequences (10000178 bp)...\n",
      "[M::process] read 108264 sequences (10000141 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 51236, 0, 22)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (145, 185, 230)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 400)\n",
      "[M::mem_pestat] mean and std.dev: (189.31, 63.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 485)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (114, 158, 263)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 561)\n",
      "[M::mem_pestat] mean and std.dev: (131.82, 63.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 710)\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 106336 reads in 7.084 CPU sec, 13.456 real sec\n",
      "[M::process] read 108866 sequences (10000007 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20, 51941, 0, 17)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (80, 107, 210)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 470)\n",
      "[M::mem_pestat] mean and std.dev: (119.00, 69.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 600)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (143, 182, 228)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 398)\n",
      "[M::mem_pestat] mean and std.dev: (187.08, 63.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 483)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (89, 176, 483)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1271)\n",
      "[M::mem_pestat] mean and std.dev: (169.14, 114.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1665)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 108264 reads in 7.133 CPU sec, 23.846 real sec\n",
      "[M::process] read 106080 sequences (10000082 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 52355, 0, 9)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (100, 144, 178)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 334)\n",
      "[M::mem_pestat] mean and std.dev: (144.00, 68.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 419)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (142, 181, 226)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 394)\n",
      "[M::mem_pestat] mean and std.dev: (185.62, 62.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 478)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_process_seqs] Processed 108866 reads in 7.073 CPU sec, 26.864 real sec\n",
      "[M::process] read 105898 sequences (10000056 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 51082, 0, 13)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (84, 104, 198)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 426)\n",
      "[M::mem_pestat] mean and std.dev: (130.17, 58.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 540)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (144, 184, 231)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 405)\n",
      "[M::mem_pestat] mean and std.dev: (189.11, 64.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 492)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (110, 130, 1006)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2798)\n",
      "[M::mem_pestat] mean and std.dev: (475.38, 646.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3694)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 106080 reads in 7.095 CPU sec, 25.090 real sec\n",
      "[M::process] read 106462 sequences (10000139 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 50847, 0, 11)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (54, 73, 174)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 414)\n",
      "[M::mem_pestat] mean and std.dev: (115.38, 97.82)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 534)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (143, 182, 227)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 395)\n",
      "[M::mem_pestat] mean and std.dev: (186.61, 62.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 479)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (83, 125, 196)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 422)\n",
      "[M::mem_pestat] mean and std.dev: (118.30, 58.38)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 535)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 105898 reads in 7.115 CPU sec, 27.827 real sec\n",
      "[M::process] read 107332 sequences (10000102 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (18, 51182, 0, 16)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (89, 126, 175)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 347)\n",
      "[M::mem_pestat] mean and std.dev: (122.25, 49.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 433)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (144, 183, 229)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 399)\n",
      "[M::mem_pestat] mean and std.dev: (188.25, 63.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 484)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (52, 105, 220)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 556)\n",
      "[M::mem_pestat] mean and std.dev: (120.33, 80.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 724)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 106462 reads in 7.239 CPU sec, 19.560 real sec\n",
      "[M::process] read 104428 sequences (10000060 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 51264, 0, 13)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (54, 90, 172)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 408)\n",
      "[M::mem_pestat] mean and std.dev: (108.73, 60.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 526)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (143, 183, 228)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 398)\n",
      "[M::mem_pestat] mean and std.dev: (187.11, 62.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 483)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (58, 108, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 715)\n",
      "[M::mem_pestat] mean and std.dev: (144.67, 98.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 934)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 107332 reads in 8.019 CPU sec, 16.702 real sec\n",
      "[M::process] read 38852 sequences (3652257 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 50337, 0, 10)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (68, 135, 213)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 503)\n",
      "[M::mem_pestat] mean and std.dev: (120.09, 68.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (143, 183, 229)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 401)\n",
      "[M::mem_pestat] mean and std.dev: (188.26, 63.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 487)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (102, 140, 152)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (2, 252)\n",
      "[M::mem_pestat] mean and std.dev: (116.11, 42.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 302)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 104428 reads in 7.304 CPU sec, 19.920 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 18569, 0, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (141, 179, 224)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 390)\n",
      "[M::mem_pestat] mean and std.dev: (184.02, 61.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 473)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 38852 reads in 2.748 CPU sec, 11.525 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem ref/GCF_000005845.2_ASM584v2_genomic.fna.gz trimmed/R1.trimmed.fastq trimmed/R2.trimmed.fastq\n",
      "[main] Real time: 186.797 sec; CPU: 61.124 sec\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir align\n",
    "bwa mem ref/GCF_000005845.2_ASM584v2_genomic.fna.gz trimmed/R1.trimmed.fastq trimmed/R2.trimmed.fastq > align/GCF_000005845.2_ASM584v2_genomic.sam"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aligned reads to the reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools view -S -b align/*.sam > align/GCF_000005845.2_ASM584v2_genomic.bam"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compressed sam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "892776 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "0 + 0 secondary\n",
      "258 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "891649 + 0 mapped (99.87% : N/A)\n",
      "892518 + 0 paired in sequencing\n",
      "446259 + 0 read1\n",
      "446259 + 0 read2\n",
      "888554 + 0 properly paired (99.56% : N/A)\n",
      "890412 + 0 with itself and mate mapped\n",
      "979 + 0 singletons (0.11% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat align/*bam # 99.87 % of reads were mapped"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Got statistics from bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools sort align/*bam -o align/GCF_000005845.2_ASM584v2_genomic.sorted.bam"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sorted bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index align/*sorted.bam"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Indexed sorted bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Viewed data in IGV. Amazing!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 3\n",
    "***24/10/2023***\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n"
     ]
    }
   ],
   "source": [
    "!samtools mpileup -f /home/glitch/bioPrac_p1/ref/GCF_000005845.2_ASM584v2_genomic.fna /home/glitch/bioPrac_p1/align/GCF_000005845.2_ASM584v2_genomic.sorted.bam > var_call/GCF_000005845.2_ASM584v2_genomic.mpileup"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made mpileup file for variance callig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: No p-value threshold provided, so p-values will not be calculated\n",
      "Min coverage:\t8\n",
      "Min reads2:\t2\n",
      "Min var freq:\t0.51\n",
      "Min avg qual:\t15\n",
      "P-value thresh:\t0.99\n",
      "Reading input from var_call/GCF_000005845.2_ASM584v2_genomic.mpileup\n",
      "4641343 bases in pileup file\n",
      "4456197 met minimum coverage of 8x\n",
      "6 SNPs predicted\n"
     ]
    }
   ],
   "source": [
    "!varscan mpileup2snp var_call/*mpileup --min-var-freq 0.4 --variants --output-vcf 1 > var_call/vs_results.vcf"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made variance calling of changes that exists in >= 51% of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-10-24 21:28:12--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz\n",
      "Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.10, 2607:f220:41e:250::13, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3421602 (3.3M) [application/x-gzip]\n",
      "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.gbff.gz’\n",
      "\n",
      "GCF_000005845.2_ASM 100%[===================>]   3.26M  2.00MB/s    in 1.6s    \n",
      "\n",
      "2023-10-24 21:28:15 (2.00 MB/s) - ‘GCF_000005845.2_ASM584v2_genomic.gbff.gz’ saved [3421602/3421602]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downloaded gebank file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"k12.genome : ecoli_K12\" > snpEff.config"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made snpEff cofig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p db/k12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "gunzip *gbff.gz\n",
    "cp *gbff db/k12/genes.gbff"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made folder for database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!snpEff build -genbank -v k12"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Made database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!snpEff ann k12 var_call/vs_results.vcf -dataDir data -config snpEff.config> var_call/vs_annotated_results.vcf"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Got annotated vcf"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Its time to make a report! ;')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioproject_1",
   "language": "python",
   "name": "bioproject_1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
